### PR TITLE
util: Remove code to cache datadir

### PR DESCRIPTION
**Pull request by maflcko**
The rationale for caching the datadir is given as

```
    // This can be called during exceptions by LogPrintf(), so we cache the
    // value so we don't have to do memory allocations after that.
```

However, since 8c2d695c4a45bdd9378c7970b0fcba6e1efc01f9, the debug log location is actually cached itself in `m_file_path`.

So remove the now-useless caching.

### CODE REVIEW

**Comment by MarcoFalke** [comment]
Will reopen after #16252

**Comment by laanwj** [comment]
Concept ACK

File: src/util/system.cpp [hunk 1]
```diff
@@ -702,23 +702,9 @@ fs::path GetDefaultDataDir()
 #endif
 }
 
-static fs::path g_blocks_path_cache_net_specific;
```
Comment (laanwj) [inline src/util/system.cpp]: happy to see this code simplified, it has been a frequent source of errors in the past

**Comment by MarcoFalke** [comment]
Rebased to aaaa6600

**Comment by laanwj** [comment]
A side-effect of caching was that this code was only run the first call to `GetDataDir`.  Right now, it will try to create the data directory every time:
```c++
    if (fs::create_directories(path)) {
        // This is the first run, create wallets subdirectory too
        fs::create_directories(path / "wallets");
    }
```
A simple solution to avoid this would be to add a static bool flag for this.

**Comment by promag** [comment]
> but also a change in behavior if, for some reason, the datadir disappears

that's good?

@laanwj wouldn't it be better to just call `create_directories` as soon as possible, like it should belong to the startup? Currently is kind of hard to know when directories are actually created.

**Comment by laanwj** [comment]
> that's good?

That definitely isn't good, if the datadir disappears while running you want the process to die as soon as possible and not mess with directories.

> @laanwj wouldn't it be better to just call create_directories  as soon as possible, like it should belong to the startup? Currently is  kind of hard to know when directories are actually created.

Yes that would be better, however this is very hard to reason about, and review, with the complexity of our startup sequence. Maybe better to leave for a future PR, which is why I suggested the most straightforward solution.

**Comment by MarcoFalke** [comment]
It is indeed weird to write to disk in a method, whose name implies it is a simple getter.

Added a guard to the disk access for now, but this should be improved in the future. (See also "Fix datadir handling" #15864)

**Comment by MarcoFalke** [comment]
I  think this is not worth it, so giving up on this for now. Someone should move out the `fs::create_directories` first.

**Pull request closed without merging**